File: src/ingest/block.ts
// src/ingest/block.ts
//
// The WASM parser's JS side: instantiate the module, hand it a block of
// whole rows, lift the active planes back out of the slab.
//
// This lives apart from worker.ts on purpose. It is the riskiest index
// arithmetic in the build -- slab plane -> cube interface, with a row stride
// of MAX_BLOCK_ROWS -- so test_ingest.mjs (T4) drives *this* code directly in
// node rather than a reimplementation of it. worker.ts is then only plumbing:
// read bytes, call this, transfer the result.

import { SLAB_METRICS } from './header';

/** block.c's MAX_BLOCK_ROWS. Compiled in, and checked against the module's
 * own `slab_rows()` at instantiation. */
export const MAX_BLOCK_ROWS = 4096;

/** block.c's ABI_VERSION. */
export const PARSER_ABI = 2;

/**
 * Marker on the block-slab overflow error.
 *
 * A worker reports failure by posting `error.message`, so the Error class is
 * gone by the time ingest sees it. This marker is how the retry recognises the
 * one failure it can actually fix -- by re-cutting the file into smaller
 * blocks -- and it must not be localised or reworded.
 */
export const OVERFLOW_MARKER = 'block-slab-overflow';

/** True for the overflow error above, wherever it has been stringified. */
export function isOverflowError(message: string): boolean {
  return message.includes(OVERFLOW_MARKER);
}

export const NEWLINE = 10;

export interface ParserExports {
  memory: WebAssembly.Memory;
  inbuf_ptr(): number;
  inbuf_size(): number;
  slab_ptr(): number;
  row_hour_ptr(): number;
  row_tou_ptr(): number;
  slab_rows(): number;
  slab_metrics(): number;
  abi_version(): number;
  last_rows(): number;
  last_overflow(): number;
  last_wide_field(): number;
  last_bad_row(): number;
  last_feb29(): number;
  last_year_mismatch(): number;
  slab_fill_nan(rows: number): void;
  parse_block(len: number, year: number): number;
}

/**
 * One parsed block, ready to scatter into a cube.
 *
 * A ROW LIST, not an hour window: row `r` holds the values at `rowHour[r]`,
 * and rows appear in the order the bytes did. Nothing downstream may assume
 * that order means anything, which is what makes a value-sorted export load
 * identically to a date-ordered one.
 */
export interface BlockPayload {
  /** Rows emitted -- valid rows only. Feb 29 and refused rows are counted
   * separately and occupy no slot. */
  rows: number;
  /** data[p * rows + r]: active plane p, row r. Plane-major, so lifting one
   * interface out of the slab stays a contiguous copy. */
  data: Float32Array;
  /** Hour of the year each row lands on, 0..8759. Length `rows`. */
  rowHour: Uint16Array;
  /** Per-row TOU code straight from the file's TOU column. Length `rows`. */
  rowTou: Uint8Array;
  /** Rows dropped because they are Feb 29 (D4) -- intended, and stated. */
  feb29: number;
}

/** A block that contributed nothing. Freshly allocated every time: the
 * buffers are transferred out of the worker, and a transferred buffer is
 * detached, so a shared constant would break the second empty block. */
function emptyPayload(feb29 = 0): BlockPayload {
  return {
    rows: 0,
    data: new Float32Array(0),
    rowHour: new Uint16Array(0),
    rowTou: new Uint8Array(0),
    feb29,
  };
}

/**
 * Instantiate the parser. Every worker needs its own instance: linear memory
 * cannot be shared without SharedArrayBuffer, which D2 forbids on GitHub
 * Pages.
 *
 * Both slab dimensions are checked against this module's mirrored constants.
 * A rebuilt block.c with a different NUM would otherwise be read with the old
 * stride -- every plane after the first shifted by a fixed offset, every
 * number plausible.
 */
export async function instantiateParser(module: WebAssembly.Module): Promise<ParserExports> {
  const instance = await WebAssembly.instantiate(module, {});
  const parser = instance.exports as unknown as ParserExports;
  // A stale block.wasm against updated TypeScript must fail HERE. The ABI
  // check catches the case the dimension checks cannot: a binary whose slab is
  // the right shape but means something else -- v1's columns were hours, v2's
  // are rows, and reading one as the other is wrong numbers, not an error.
  if (typeof parser.abi_version !== 'function' || parser.abi_version() !== PARSER_ABI) {
    const found = typeof parser.abi_version === 'function' ? parser.abi_version() : 'none';
    throw new Error(
      `block.wasm reports ABI ${found}, but this build speaks ABI ${PARSER_ABI}. ` +
        `Rebuild the parser: parser/build.sh.`,
    );
  }
  if (parser.slab_rows() !== MAX_BLOCK_ROWS || parser.slab_metrics() !== SLAB_METRICS) {
    throw new Error(
      `block.wasm reports slab_rows()=${parser.slab_rows()}, ` +
        `slab_metrics()=${parser.slab_metrics()}, but this build mirrors ` +
        `MAX_BLOCK_ROWS=${MAX_BLOCK_ROWS}, SLAB_METRICS=${SLAB_METRICS}. ` +
        `Rebuild the parser or fix the constants.`,
    );
  }
  return parser;
}

/**
 * Parse `bytes[from, to)` -- which must start at a row boundary and end just
 * after a `\n` -- and lift `activePlanes` out of the slab.
 *
 * Every row carries its own hour, so the result is position-independent twice
 * over: any block can be parsed by any worker in any order (D10), and the rows
 * inside a block need not be in any order either.
 *
 * `year` is the case's year, from the file's first data row. Rows carrying a
 * different one are refused rather than folded onto the same 8,760 hours.
 */
export function parseBytes(
  parser: ParserExports,
  bytes: Uint8Array,
  from: number,
  to: number,
  activePlanes: Int32Array,
  year: number,
): BlockPayload {
  const length = to - from;
  if (length <= 0) return emptyPayload();

  const inbufPtr = parser.inbuf_ptr();
  const inbufSize = parser.inbuf_size();
  if (length + 1 > inbufSize) {
    throw new Error(`Block of ${length} B exceeds the ${inbufSize} B WASM input window.`);
  }

  const memory = new Uint8Array(parser.memory.buffer);
  memory.set(bytes.subarray(from, to), inbufPtr);
  let padded = length;
  // parse_block only emits a row once it sees the terminator, so a file that
  // does not end in a newline would otherwise lose its final row.
  if (memory[inbufPtr + padded - 1] !== NEWLINE) memory[inbufPtr + padded++] = NEWLINE;

  // Clear the whole slab, not only the rows this block will fill. A stale
  // float left over from the previous block is a plausible-looking wrong
  // number, which is precisely the failure this project cannot see by eye.
  parser.slab_fill_nan(MAX_BLOCK_ROWS);
  const rows = parser.parse_block(padded, year);

  // Counters are read BEFORE the empty-block shortcut: a block that emitted no
  // rows because every row was refused has the most to report, and returning
  // early on rows === 0 would swallow exactly that.
  const wide = parser.last_wide_field();
  if (wide > 0) {
    throw new Error(
      `${wide} field(s) sit past column ${SLAB_METRICS + 3} — this export is wider than ` +
        `parser/block.c's ${SLAB_METRICS}-interface slab. Those columns would be dropped ` +
        `silently, so the load is refused instead. Widen NUM in block.c and rebuild.`,
    );
  }
  const bad = parser.last_bad_row();
  if (bad > 0) {
    throw new Error(
      `${bad} row(s) carry a Date or Hour this parser could not read (expected M/D/YYYY and ` +
        `an hour-ending 1-24). Those rows would be dropped silently, so the load is refused.`,
    );
  }
  const mismatched = parser.last_year_mismatch();
  if (mismatched > 0) {
    throw new Error(
      `${mismatched} row(s) carry a year other than ${year}, which is the year on this file's ` +
        `first data row. A case is one calendar year of 8,760 hours, so a second year would be ` +
        `folded onto the same hours. Split the export by year and load the files separately.`,
    );
  }
  const overflow = parser.last_overflow();
  if (overflow > 0) {
    throw new Error(
      `${OVERFLOW_MARKER}: ${overflow} row(s) past the ${MAX_BLOCK_ROWS}-row block slab — this ` +
        `block held more rows than its byte length predicted.`,
    );
  }

  const feb29 = parser.last_feb29();
  if (rows === 0) return emptyPayload(feb29);

  // Plane-major on both sides, so each interface is one contiguous copy.
  const slab = new Float32Array(
    parser.memory.buffer,
    parser.slab_ptr(),
    SLAB_METRICS * MAX_BLOCK_ROWS,
  );
  const data = new Float32Array(activePlanes.length * rows);
  for (let p = 0; p < activePlanes.length; p++) {
    const src = activePlanes[p] * MAX_BLOCK_ROWS;
    data.set(slab.subarray(src, src + rows), p * rows);
  }

  const rowHour = new Uint16Array(parser.memory.buffer, parser.row_hour_ptr(), MAX_BLOCK_ROWS);
  const rowTou = new Uint8Array(parser.memory.buffer, parser.row_tou_ptr(), MAX_BLOCK_ROWS);

  return {
    rows,
    data,
    // Copied out of linear memory, not viewed into it: these are transferred
    // to the main thread, and the next block reuses the same addresses.
    rowHour: rowHour.slice(0, rows),
    rowTou: rowTou.slice(0, rows),
    feb29,
  };
}

/** Index just past the first `\n` at or after `from`, or -1. */
export function afterNextNewline(bytes: Uint8Array, from: number): number {
  const at = bytes.indexOf(NEWLINE, Math.max(0, from));
  return at < 0 ? -1 : at + 1;
}

File: parser/block.c
// The WASM CSV parser for GridView MONITORED-INTERFACE exports.
//
// Shape of the input:
//
//   line 1   Interface Hourly 'Power Flow (MW)' Data for Year 2034
//   line 2   (blank)
//   line 3   (From the first hour of 1/1/2034 ... Column identifier -- InterfaceName)
//   line 4   (blank)
//   line 5   Date, Hour, TOU,<interface 1>,<interface 2>,...
//   line 6+  one row per HOUR
//
// The four preamble lines and the header are consumed on the JS side; this
// module only ever sees whole DATA rows. Four invariants:
//
// 1. MEMORY. Every Worker needs its own WASM instance, so per-instance memory
//    must be a function of BLOCK size, never of case size: 12 MiB input
//    window + an 8 MiB slab.
// 2. BLOCK INDEPENDENCE. Each row's hour comes from its OWN Date and
//    Hour fields, never from a running row counter, so a worker needs its
//    bytes and not its position and blocks may be parsed in any order.
// 3. ONE ROW = ONE HOUR. Every interface is a column, so a row is a complete
//    hour and no hour is ever split across two blocks.
// 4. ROW ORDER DOES NOT MATTER. Output is a ROW LIST: slab column `r` holds
//    row `r`'s values and rowHour[r] says which hour of the year that row
//    lands on. The main thread scatters it. Nothing here assumes rows arrive
//    sorted, inside a block or across blocks.
//
// (4) replaced an hour-windowed slab indexed by `hour - baseHour`, where
// baseHour came from the block's FIRST row. Rows out of order within a block
// were placed correctly, but any row EARLIER than its block's first row fell
// outside the window and the whole load was refused -- so a value-sorted or
// descending export could not be loaded at all, and the error blamed block
// sizing. The slab is still [NUM x 4096] and still 8 MiB; only what a column
// means has changed.
//
// Build: ./build.sh

#include <wasm_simd128.h>

// Interface columns per export. The reference exports carry 167; 512 leaves
// room for a wider study without another rebuild, and costs 8 MiB of slab at
// MAX_BLOCK_ROWS. A file with more columns than this is REFUSED on the JS
// side (last_wide_field), never silently truncated.
#define NUM             512
#define KEY_COLS        3           // Date, Hour, TOU
#define BLOCK_BYTES     (12u * 1024u * 1024u)
// Rows one block may hold. One row is one hour, so this is the same 8 MiB the
// hour-windowed slab cost. Blocks are sized on the JS side from the file's own
// row length so this is never reached; reaching it is counted and refused
// rather than wrapped.
#define MAX_BLOCK_ROWS  4096u

// Bumped on any change to the exported surface or the slab's meaning.
// block.wasm is committed, so a stale binary against updated TypeScript has to
// fail at instantiate rather than produce a wrong number at parse time. v2 is
// the row list; v1 was the hour-windowed slab.
#define ABI_VERSION     2u

static unsigned char inbuf[BLOCK_BYTES];
// slab[m * MAX_BLOCK_ROWS + r] -- PLANE-major, so lifting one interface's
// values for the whole block stays a contiguous copy on the JS side. Row-major
// would make that a strided gather for no gain: the scatter into the cube is
// per cell either way.
static float         slab[(unsigned)NUM * MAX_BLOCK_ROWS];

// Where each emitted row lands: hour of the year, 0..8759. Every row carries
// its own placement, which is the whole of invariant 4.
static unsigned short rowHour[MAX_BLOCK_ROWS];

// Per-ROW TOU code, 0 = OffPeak, 1 = OnPeak. TOU is FILE DATA and is never
// recomputed from the calendar (footgun 17) -- the real rule is HE 6-22
// Mon-Sat with Sunday fully OffPeak, varies by tariff, and drops holidays. It
// is read here rather than in a second JS pass purely so the block's bytes are
// scanned once; the value itself still comes from the file's own TOU column.
static unsigned char rowTou[MAX_BLOCK_ROWS];

static unsigned g_rows, g_overflow, g_wideField, g_badRow, g_feb29, g_yearMismatch;

__attribute__((export_name("inbuf_ptr")))     unsigned char* inbuf_ptr(void)    { return inbuf; }
__attribute__((export_name("inbuf_size")))    unsigned       inbuf_size(void)   { return BLOCK_BYTES; }
__attribute__((export_name("slab_ptr")))      float*         slab_ptr(void)     { return slab; }
__attribute__((export_name("row_hour_ptr")))  unsigned short* row_hour_ptr(void){ return rowHour; }
__attribute__((export_name("row_tou_ptr")))   unsigned char* row_tou_ptr(void)  { return rowTou; }
__attribute__((export_name("slab_rows")))     unsigned       slab_rows(void)    { return MAX_BLOCK_ROWS; }
__attribute__((export_name("slab_metrics")))  unsigned       slab_metrics(void) { return NUM; }
__attribute__((export_name("abi_version")))   unsigned       abi_version(void)  { return ABI_VERSION; }
__attribute__((export_name("last_rows")))     unsigned       last_rows(void)    { return g_rows; }
// Rows past MAX_BLOCK_ROWS: the block held more rows than the slab has
// columns. JS sizes blocks to make this unreachable and shrinks them if it
// happens anyway, so this is a backstop, never a normal outcome.
__attribute__((export_name("last_overflow")))   unsigned     last_overflow(void){ return g_overflow; }
// Fields past the slab's NUM columns, and rows whose Date/Hour did not
// resolve to an hour of the year. Both are silent data loss if unreported,
// which is the one failure mode this project cannot see by eye. Feb 29 is
// counted apart from them because dropping it is intended and stated, not a
// fault.
__attribute__((export_name("last_wide_field"))) unsigned     last_wide_field(void){ return g_wideField; }
__attribute__((export_name("last_bad_row")))    unsigned     last_bad_row(void)  { return g_badRow; }
__attribute__((export_name("last_feb29")))      unsigned     last_feb29(void)    { return g_feb29; }
// Rows whose Date carries a year other than the one JS read from the file's
// first data row. date_to_day deliberately ignores the year -- it maps M/D to
// a day of THE case's year -- so without this a file holding two years would
// fold both onto the same 8,760 hours. With rows in file order that shows up
// as duplicate hours; shuffled, it need not. Counted here, refused by JS.
__attribute__((export_name("last_year_mismatch"))) unsigned  last_year_mismatch(void){ return g_yearMismatch; }

// Cumulative days before each month, non-leap. Feb 29 is dropped at ingest
// so a leap year uses the same table and the Feb 29 rows are skipped.
static const unsigned short CUM[12] = {0,31,59,90,120,151,181,212,243,273,304,334};

// f64 digit accumulation -- measured FASTER than the Eisel-Lemire u64-mantissa
// approach on wasm32 in lever-1 (130.2 ms vs 157.0 ms). Do NOT "optimize" this
// into fast_float's algorithm; it loses here.
// Exponent notation IS present in real exports -- the interface exports carry
// `2.568664E-03` and `1.338663E-04` for near-zero flows and costs, and the
// export writes tiny magnitudes that way rather than as long decimal strings.
// Returning NaN on them (as this did until T4 caught it cell-by-cell) is
// silent data loss: the cell simply reads as absent, which every kernel is
// designed to skip without complaint.
static inline float parse_float(const unsigned char* p, const unsigned char* e) {
  if (e <= p) return 0.0f / 0.0f;
  int neg = 0;
  if (*p == '-') { neg = 1; p++; }
  else if (*p == '+') { p++; }

  // Integer part, then `frac / scale`. This rounds twice where a correctly
  // rounded strtod rounds once, and on the widest columns that shows up as a
  // 1-ulp float32 difference on a handful of cells.
  //
  // MEASURED, do not "fix" this:
  //   * Folding both parts into one f64 mantissa with a single division
  //     costs 20% throughput (204 -> 160 MiB/s) and changes none of the
  //     differing cells -- those values carry 18 significant digits, so the
  //     mantissa is inexact either way.
  //   * Eisel-Lemire is already refuted for wasm32 here.
  // The residual is below f32 storage precision by construction, which is
  // the only reason it is acceptable. T4 gates it at <= 1 ulp, not at zero.
  double ip = 0.0;
  while (p < e) {
    unsigned d = (unsigned)(*p - '0');
    if (d > 9) break;
    ip = ip * 10.0 + (double)d;
    p++;
  }
  double v = ip;
  if (p < e && *p == '.') {
    p++;
    double f = 0.0, sc = 1.0;
    while (p < e) {
      unsigned d = (unsigned)(*p - '0');
      if (d > 9) break;
      f = f * 10.0 + (double)d;
      sc *= 10.0;
      p++;
    }
    v += f / sc;
  }
  if (p < e && (*p == 'e' || *p == 'E')) {
    p++;
    int eneg = 0;
    if (p < e && (*p == '-' || *p == '+')) { eneg = (*p == '-'); p++; }
    int ex = 0;
    while (p < e) {
      unsigned d = (unsigned)(*p - '0');
      if (d > 9) break;
      ex = ex * 10 + (int)d;
      if (ex > 400) ex = 400;   // past f32 range either way
      p++;
    }
    // Exponentiation by squaring: no libm in a freestanding module, and a
    // POW10 table would still need this path for the out-of-table cases.
    double scale = 1.0, base = 10.0;
    for (int k = ex; k; k >>= 1) { if (k & 1) scale *= base; base *= base; }
    v = eneg ? v / scale : v * scale;
  }

  // Anything left over is garbage, not a number. Kept strict: the old loop
  // bailed to NaN on the first non-digit, and a partial parse of a mangled
  // field would be a plausible-looking wrong number.
  if (p != e) return 0.0f / 0.0f;
  return (float)(neg ? -v : v);
}

static inline unsigned parse_uint(const unsigned char* p, const unsigned char* e) {
  unsigned v = 0;
  for (; p < e; p++) {
    unsigned d = (unsigned)(*p - '0');
    if (d > 9) break;
    v = v * 10u + d;
  }
  return v;
}

// M/D/YYYY -> day-of-year, with the YEAR handed back through `yearOut` so the
// caller can check every row belongs to the same year. Two distinct failure
// values: FEB29 for the day this build drops on purpose (D4), NO_DAY for a
// date it could not read. They are counted separately so "the leap day went"
// never looks like "the file is mangled", and neither looks like a clean
// parse.
#define FEB29  0xFFFFFFFEu
#define NO_DAY 0xFFFFFFFFu
static inline unsigned date_to_day(const unsigned char* p, const unsigned char* e,
                                   unsigned* yearOut) {
  unsigned month = 0, day = 0, year = 0;
  while (p < e && *p != '/') { month = month * 10u + (unsigned)(*p - '0'); p++; }
  p++;
  while (p < e && *p != '/') { day = day * 10u + (unsigned)(*p - '0'); p++; }
  p++;
  while (p < e) {
    unsigned d = (unsigned)(*p - '0');
    if (d > 9) break;
    year = year * 10u + d;
    p++;
  }
  *yearOut = year;
  if (month < 1 || month > 12 || day < 1 || day > 31) return NO_DAY;
  if (month == 2 && day == 29) return FEB29;   // D4: Feb 29 dropped
  return CUM[month - 1] + day - 1;
}

/**
 * NaN-fill the value columns for `rows` rows.
 *
 * Still necessary with a row list: a row carrying fewer fields than the header
 * leaves the planes past its last field untouched, and a stale float from the
 * previous block is a plausible-looking wrong number -- precisely the failure
 * this project cannot see by eye. The fill is plane-strided because the slab
 * is plane-major.
 */
__attribute__((export_name("slab_fill_nan")))
void slab_fill_nan(unsigned rows) {
  if (rows > MAX_BLOCK_ROWS) rows = MAX_BLOCK_ROWS;
  const float nan = 0.0f / 0.0f;
  for (unsigned m = 0; m < (unsigned)NUM; m++) {
    float* column = slab + m * MAX_BLOCK_ROWS;
    for (unsigned r = 0; r < rows; r++) column[r] = nan;
  }
}

// Parse a block of WHOLE rows. `len` bytes in inbuf, starting at a row boundary
// and ending just after a '\n'. `year` is the case's year, read by JS from the
// file's first data row; every row's own year is checked against it.
//
// Slab layout: [interface * MAX_BLOCK_ROWS + row], with rowHour[row] carrying
// where that row belongs. Rows are emitted in the order they appear in the
// bytes, which is not necessarily hour order and does not need to be.
//
// Interfaces are positional here -- source column `c` writes plane `c -
// KEY_COLS`. Production maps a plane to its cube index by TRIMMED HEADER NAME
// (footgun 18) through a JS-side plan array; different exports list different
// interfaces in different orders, so position is only ever a transport detail.
__attribute__((export_name("parse_block")))
unsigned parse_block(unsigned len, unsigned year) {
  const unsigned char* b = inbuf;
  unsigned col = 0, fs = 0;
  unsigned rowDay = NO_DAY, rowYear = 0;
  // Slot this row will occupy, or NO_DAY while the row is being rejected.
  unsigned slot = NO_DAY;
  g_rows = 0; g_overflow = 0; g_wideField = 0; g_badRow = 0; g_feb29 = 0;
  g_yearMismatch = 0;

  const v128_t vcomma = wasm_i8x16_splat(',');
  const v128_t vnl    = wasm_i8x16_splat('\n');
  unsigned i = 0;

  #define FIELD(END)                                                            \
    {                                                                           \
      unsigned e = (END);                                                       \
      if (e > fs && b[e - 1] == '\r') e--;                                      \
      if (col == 0) {                                                           \
        rowDay = date_to_day(b + fs, b + e, &rowYear);                          \
      } else if (col == 1) {                                                    \
        unsigned hourOfDay = parse_uint(b + fs, b + e);                         \
        slot = NO_DAY;                                                          \
        if (rowDay == FEB29) {                                                  \
          g_feb29++;                     /* dropped on purpose (D4) */          \
        } else if (rowDay == NO_DAY || hourOfDay < 1 || hourOfDay > 24) {       \
          g_badRow++;                    /* unreadable Date or Hour */          \
        } else if (rowYear != year) {                                           \
          g_yearMismatch++;              /* a second year in one case */        \
        } else if (g_rows >= MAX_BLOCK_ROWS) {                                  \
          g_overflow++;                  /* more rows than the slab holds */    \
        } else {                                                                \
          slot = g_rows++;                                                      \
          rowHour[slot] = (unsigned short)(rowDay * 24u + (hourOfDay - 1u));    \
          rowTou[slot] = 0;                                                     \
        }                                                                       \
      } else if (col == 2) {                                                    \
        /* "OnPeak" / "OffPeak" differ at byte 1: 'n' vs 'f'. */                \
        if (slot != NO_DAY) {                                                   \
          rowTou[slot] = (unsigned char)((e > fs + 1 && b[fs + 1] == 'n') ? 1 : 0); \
        }                                                                       \
      } else {                                                                  \
        unsigned m = col - (unsigned)KEY_COLS;                                  \
        if (m >= (unsigned)NUM) g_wideField++;                                  \
        else if (slot != NO_DAY) {                                              \
          slab[m * MAX_BLOCK_ROWS + slot] = parse_float(b + fs, b + e);         \
        }                                                                       \
      }                                                                         \
    }

  for (; i + 16 <= len; i += 16) {
    v128_t chunk = wasm_v128_load(b + i);
    v128_t hit   = wasm_v128_or(wasm_i8x16_eq(chunk, vcomma), wasm_i8x16_eq(chunk, vnl));
    unsigned mask = (unsigned)wasm_i8x16_bitmask(hit);
    while (mask) {
      unsigned pos = i + (unsigned)__builtin_ctz(mask);
      mask &= mask - 1;
      FIELD(pos)
      fs = pos + 1;
      col++;
      /* g_rows counts EMITTED rows and is advanced when a slot is claimed, so
         a newline only resets the per-row state. */
      if (b[pos] == '\n') { col = 0; rowDay = NO_DAY; slot = NO_DAY; }
    }
  }
  for (; i < len; i++) {
    unsigned char c = b[i];
    if (c == ',' || c == '\n') {
      FIELD(i)
      fs = i + 1;
      col++;
      if (c == '\n') { col = 0; rowDay = NO_DAY; slot = NO_DAY; }
    }
  }
  #undef FIELD
  return g_rows;
}
